% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/quantgen.R
\name{plotHistMinAllelFreq}
\alias{plotHistMinAllelFreq}
\title{Minor allele frequencies}
\usage{
plotHistMinAllelFreq(
  X = NULL,
  mafs = NULL,
  main = NULL,
  xlim = c(0, 0.5),
  col = "grey",
  border = "white",
  las = 1,
  breaks = "FD",
  add.ml.beta = FALSE,
  verbose = 1,
  ...
)
}
\arguments{
\item{X}{matrix of bi-allelic SNP genotypes encoded, for each SNP, in number of copies of its minor allele, i.e. as allele doses in [0,2], with genotypes in rows and SNPs in columns; if X is not NULL, minor allele frequencies will be estimated via \code{\link{estimSnpAf}}}

\item{mafs}{vector of minor allele frequencies (not used if X is not NULL)}

\item{main}{string for the main title}

\item{xlim}{limits of the x-axis}

\item{col}{color for the bars}

\item{border}{color for the border of the bars}

\item{las}{see ?par}

\item{breaks}{see ?hist}

\item{add.ml.beta}{if TRUE, add a curve corresponding to a Beta distribution with parameters inferred by maximum likelihood}

\item{verbose}{verbosity level (0/1)}

\item{...}{arguments to be passed to hist()}
}
\value{
invisible list with the output of \code{hist} and \code{optim} (if \code{add.ml.beta=TRUE})
}
\description{
Plot the histogram of the minor allele frequency per SNP.
}
\author{
Timothee Flutre
}
