% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bioinfo_sequences.R
\name{rngVcf2df}
\alias{rngVcf2df}
\title{Convert ranges to data.frame}
\usage{
rngVcf2df(
  vcf,
  with.coords = TRUE,
  with.alleles = TRUE,
  single.ref = FALSE,
  single.alt = FALSE
)
}
\arguments{
\item{vcf}{CollapsedVCF (see pkg \href{http://bioconductor.org/packages/VariantAnnotation/}{VariantAnnotation})}

\item{with.coords}{if TRUE, the output will contain variant coordinates}

\item{with.alleles}{if TRUE, the output will contain variant alleles}

\item{single.ref}{if TRUE, only records with a single 'ref' are kept}

\item{single.alt}{if TRUE, only records with a single 'alt' are kept}
}
\value{
data.frame with variants in rows
}
\description{
Non-bi-allelic variants are discarded.
}
\seealso{
\code{\link{subsetVcfOnAllelicity}}
}
\author{
Timothee Flutre
}
