% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bioinfo_sequences.R
\name{setGt2Na}
\alias{setGt2Na}
\title{Set GT to NA}
\usage{
setGt2Na(
  vcf.file,
  genome = "",
  out.file,
  yieldSize = NA_integer_,
  dict.file = NULL,
  seq.id = NULL,
  seq.start = NULL,
  seq.end = NULL,
  min.gq = 90,
  verbose = 1
)
}
\arguments{
\item{vcf.file}{path to the VCF file (if the bgzip index doesn't exist in the same directory, it will be created)}

\item{genome}{genome identifier (e.g. "VITVI_12x2")}

\item{out.file}{path to the output VCF file (a bgzip index will be created in the same directory)}

\item{yieldSize}{number of records to yield each time the file is read from (see ?TabixFile) if seq.id is NULL}

\item{dict.file}{path to the SAM dict file (see \url{https://broadinstitute.github.io/picard/command-line-overview.html#CreateSequenceDictionary}) if seq.id is specified with no start/end}

\item{seq.id}{sequence identifier to work on (e.g. "chr2")}

\item{seq.start}{start of the sequence to work on (if NULL, whole seq)}

\item{seq.end}{end of the sequence to work on (if NULL, whole seq)}

\item{min.gq}{minimum GQ below which GT is set to missing}

\item{verbose}{verbosity level (0/1)}
}
\value{
the destination file path as an invisible character(1)
}
\description{
Set genotypes (GT field) to missing if the genotype quality (GQ field) isn't high enough.
}
\seealso{
\code{\link{filterVariantCalls}}
}
\author{
Timothee Flutre
}
