% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bioinfo_sequences.R
\name{vcf2genoClasses}
\alias{vcf2genoClasses}
\title{Convert VCF to genotypic classes}
\usage{
vcf2genoClasses(
  vcf.file,
  genome = "",
  gclasses.file,
  ca.file,
  yieldSize = NA_integer_,
  dict.file = NULL,
  seq.id = NULL,
  seq.start = NULL,
  seq.end = NULL,
  na.string = NA,
  single.alt = TRUE,
  verbose = 1
)
}
\arguments{
\item{vcf.file}{path to the VCF file (if the bgzip index doesn't exist in the same directory, it will be created)}

\item{genome}{genome identifier (e.g. "VITVI_12x2")}

\item{gclasses.file}{path to the output file to record genotypes into genotypic classes (will be gzipped if suffix is ".gz")}

\item{ca.file}{path to the output file to record SNP 1-based coordinates and alleles (will be gzipped if suffix is ".gz")}

\item{yieldSize}{number of records to yield each time the file is read from (see \code{?TabixFile}) if seq.id is NULL}

\item{dict.file}{path to the SAM dict file (see \url{https://broadinstitute.github.io/picard/command-line-overview.html#CreateSequenceDictionary}) if seq.id is specified with no start/end}

\item{seq.id}{see \code{\link{seqIdStartEnd2GRanges}}}

\item{seq.start}{see \code{\link{seqIdStartEnd2GRanges}}}

\item{seq.end}{see \code{\link{seqIdStartEnd2GRanges}}}

\item{na.string}{a symbol to indicate missing genotypes (e.g. NA, "NN", "--", etc)}

\item{single.alt}{if TRUE, only records with a single 'alt' are kept}

\item{verbose}{verbosity level (0/1)}
}
\value{
invisible vector with the path to the output file
}
\description{
Convert genotypes at SNPs from a VCF file into genotypic classes.
}
\seealso{
\code{\link{gtVcf2genoClasses}}, \code{\link{filterVariantCalls}}
}
\author{
Gautier Sarah, Timothee Flutre
}
