% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/bioinfo_sequences.R
\name{vcf2dosage}
\alias{vcf2dosage}
\title{Convert VCF to dose}
\usage{
vcf2dosage(
  vcf.file,
  genome = "",
  gdose.file,
  ca.file,
  yieldSize = NA_integer_,
  dict.file = NULL,
  seq.id = NULL,
  seq.start = NULL,
  seq.end = NULL,
  field = "GT",
  verbose = 1
)
}
\arguments{
\item{vcf.file}{path to the VCF file (if the bgzip index doesn't exist in the same directory, it will be created)}

\item{genome}{genome identifier (e.g. "VITVI_12x2")}

\item{gdose.file}{path to the output file to record genotypes as allele doses (will be gzipped); variants will be in rows and samples in columns}

\item{ca.file}{path to the output file to record SNP 1-based coordinates and alleles (will be gzipped)}

\item{yieldSize}{number of records to yield each time the file is read from (see ?TabixFile) if seq.id is NULL}

\item{dict.file}{path to the SAM dict file (see \url{https://broadinstitute.github.io/picard/command-line-overview.html#CreateSequenceDictionary}) if seq.id is specified with no start/end}

\item{seq.id}{see \code{\link{seqIdStartEnd2GRanges}}}

\item{seq.start}{see \code{\link{seqIdStartEnd2GRanges}}}

\item{seq.end}{see \code{\link{seqIdStartEnd2GRanges}}}

\item{field}{the genotypes to convert should come from the "GT" or "DS" fields}

\item{verbose}{verbosity level (0/1)}
}
\value{
an invisible list with both output file paths
}
\description{
Convert genotypes at bi-allelic variants from a VCF file into allele doses.
This is also doable with \code{bcftools +dosage}.
}
\seealso{
\code{\link{gtVcf2dose}}, \code{\link{dsVcf2dose}}, \code{\link{filterVariantCalls}}
}
\author{
Timothee Flutre
}
